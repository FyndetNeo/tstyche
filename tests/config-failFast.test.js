import { afterEach, describe, expect, test } from "@jest/globals";
import { clearFixture, getFixtureUrl, writeFixture } from "./__utils__/fixtureFactory.js";
import { normalizeOutput } from "./__utils__/normalizeOutput.js";
import { spawnTyche } from "./__utils__/spawnTyche.js";

const isStringTestText = `import { expect, test } from "tstyche";
test("is string?", () => {
  expect<string>().type.not.toBeString();
  expect<string>().type.not.toBeString();
});
`;

const isNumberTestText = `import { expect, test } from "tstyche";
test("is number?", () => {
  expect<number>().type.not.toBeNumber();
  expect<number>().type.not.toBeNumber();
});
`;

const fixtureUrl = getFixtureUrl("config-failFast", { generated: true });

afterEach(async () => {
  await clearFixture(fixtureUrl);
});

describe("'--failFast' command line option", () => {
  test("stops running tests after the first failure", async () => {
    await writeFixture(fixtureUrl, {
      ["__typetests__/isNumber.tst.ts"]: isNumberTestText,
      ["__typetests__/isString.tst.ts"]: isStringTestText,
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl, ["--failFast"]);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toMatchSnapshot("stderr");

    expect(exitCode).toBe(1);
  });

  test("when 'true' is passed as an argument", async () => {
    await writeFixture(fixtureUrl, {
      ["__typetests__/isNumber.tst.ts"]: isNumberTestText,
      ["__typetests__/isString.tst.ts"]: isStringTestText,
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl, ["--failFast", "true"]);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toMatchSnapshot("stderr");

    expect(exitCode).toBe(1);
  });

  test("when 'false' is passed as an argument", async () => {
    await writeFixture(fixtureUrl, {
      ["__typetests__/isNumber.tst.ts"]: isNumberTestText,
      ["__typetests__/isString.tst.ts"]: isStringTestText,
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl, ["--failFast", "false"]);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toMatchSnapshot("stderr");

    expect(exitCode).toBe(1);
  });

  test("when the option is specified several times", async () => {
    await writeFixture(fixtureUrl, {
      ["__typetests__/isNumber.tst.ts"]: isNumberTestText,
      ["__typetests__/isString.tst.ts"]: isStringTestText,
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl, [
      "--failFast",
      "isNumber",
      "--failFast",
      "false",
    ]);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toMatchSnapshot("stderr");

    expect(exitCode).toBe(1);
  });

  test("when search string is specified before the option", async () => {
    await writeFixture(fixtureUrl, {
      ["__typetests__/isNumber.tst.ts"]: isNumberTestText,
      ["__typetests__/isString.tst.ts"]: isStringTestText,
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl, ["isNumber", "--failFast"]);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toMatchSnapshot("stderr");

    expect(exitCode).toBe(1);
  });

  test("when search string is specified after the option", async () => {
    await writeFixture(fixtureUrl, {
      ["__typetests__/isNumber.tst.ts"]: isNumberTestText,
      ["__typetests__/isString.tst.ts"]: isStringTestText,
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl, ["--failFast", "isString"]);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toMatchSnapshot("stderr");

    expect(exitCode).toBe(1);
  });

  test("overrides configuration file option, when it is set to 'true'", async () => {
    const config = {
      failFast: true,
    };

    await writeFixture(fixtureUrl, {
      ["__typetests__/isNumber.tst.ts"]: isNumberTestText,
      ["__typetests__/isString.tst.ts"]: isStringTestText,
      ["tstyche.config.json"]: JSON.stringify(config, null, 2),
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl, ["--failFast", "false"]);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toMatchSnapshot("stderr");

    expect(exitCode).toBe(1);
  });

  test("overrides configuration file option, when it is set to 'false'", async () => {
    const config = {
      failFast: false,
    };

    await writeFixture(fixtureUrl, {
      ["__typetests__/isNumber.tst.ts"]: isNumberTestText,
      ["__typetests__/isString.tst.ts"]: isStringTestText,
      ["tstyche.config.json"]: JSON.stringify(config, null, 2),
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl, ["--failFast"]);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toMatchSnapshot("stderr");

    expect(exitCode).toBe(1);
  });
});

describe("'failFast' configuration file option", () => {
  test("stops running tests after the first failure", async () => {
    const config = {
      failFast: true,
    };

    await writeFixture(fixtureUrl, {
      ["__typetests__/isNumber.tst.ts"]: isNumberTestText,
      ["__typetests__/isString.tst.ts"]: isStringTestText,
      ["tstyche.config.json"]: JSON.stringify(config, null, 2),
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toMatchSnapshot("stderr");

    expect(exitCode).toBe(1);
  });
});
