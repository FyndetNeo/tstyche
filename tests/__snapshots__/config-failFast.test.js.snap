// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`'--failFast' command line option overrides configuration file option, when it is set to 'false': stderr 1`] = `
"Error: The source type is 'number'.

  1 | import { expect, test } from "tstyche";
  2 | test("is number?", () => {
> 3 |   expect<number>().type.not.toBeNumber();
    |                             ^
  4 |   expect<number>().type.not.toBeNumber();
  5 | });
  6 | 

      at ./__typetests__/isNumber.tst.ts:3:29 ❭ is number?

"
`;

exports[`'--failFast' command line option overrides configuration file option, when it is set to 'false': stdout 1`] = `
"uses TypeScript <<version>>

fail ./__typetests__/isNumber.tst.ts

Targets:    1 failed, 1 total
Test files: 1 failed, 1 total
Tests:      1 failed, 1 total
Assertions: 1 failed, 1 total
Duration:   <<timestamp>>

Ran all test files.
"
`;

exports[`'--failFast' command line option overrides configuration file option, when it is set to 'true': stderr 1`] = `
"Error: The source type is 'number'.

  1 | import { expect, test } from "tstyche";
  2 | test("is number?", () => {
> 3 |   expect<number>().type.not.toBeNumber();
    |                             ^
  4 |   expect<number>().type.not.toBeNumber();
  5 | });
  6 | 

      at ./__typetests__/isNumber.tst.ts:3:29 ❭ is number?

Error: The source type is 'number'.

  2 | test("is number?", () => {
  3 |   expect<number>().type.not.toBeNumber();
> 4 |   expect<number>().type.not.toBeNumber();
    |                             ^
  5 | });
  6 | 

      at ./__typetests__/isNumber.tst.ts:4:29 ❭ is number?

Error: The source type is 'string'.

  1 | import { expect, test } from "tstyche";
  2 | test("is string?", () => {
> 3 |   expect<string>().type.not.toBeString();
    |                             ^
  4 |   expect<string>().type.not.toBeString();
  5 | });
  6 | 

      at ./__typetests__/isString.tst.ts:3:29 ❭ is string?

Error: The source type is 'string'.

  2 | test("is string?", () => {
  3 |   expect<string>().type.not.toBeString();
> 4 |   expect<string>().type.not.toBeString();
    |                             ^
  5 | });
  6 | 

      at ./__typetests__/isString.tst.ts:4:29 ❭ is string?

"
`;

exports[`'--failFast' command line option overrides configuration file option, when it is set to 'true': stdout 1`] = `
"uses TypeScript <<version>>

fail ./__typetests__/isNumber.tst.ts

fail ./__typetests__/isString.tst.ts

Targets:    1 failed, 1 total
Test files: 2 failed, 2 total
Tests:      2 failed, 2 total
Assertions: 4 failed, 4 total
Duration:   <<timestamp>>

Ran all test files.
"
`;

exports[`'--failFast' command line option stops running tests after the first failure: stderr 1`] = `
"Error: The source type is 'number'.

  1 | import { expect, test } from "tstyche";
  2 | test("is number?", () => {
> 3 |   expect<number>().type.not.toBeNumber();
    |                             ^
  4 |   expect<number>().type.not.toBeNumber();
  5 | });
  6 | 

      at ./__typetests__/isNumber.tst.ts:3:29 ❭ is number?

"
`;

exports[`'--failFast' command line option stops running tests after the first failure: stdout 1`] = `
"uses TypeScript <<version>>

fail ./__typetests__/isNumber.tst.ts

Targets:    1 failed, 1 total
Test files: 1 failed, 1 total
Tests:      1 failed, 1 total
Assertions: 1 failed, 1 total
Duration:   <<timestamp>>

Ran all test files.
"
`;

exports[`'--failFast' command line option when 'false' is passed as an argument: stderr 1`] = `
"Error: The source type is 'number'.

  1 | import { expect, test } from "tstyche";
  2 | test("is number?", () => {
> 3 |   expect<number>().type.not.toBeNumber();
    |                             ^
  4 |   expect<number>().type.not.toBeNumber();
  5 | });
  6 | 

      at ./__typetests__/isNumber.tst.ts:3:29 ❭ is number?

Error: The source type is 'number'.

  2 | test("is number?", () => {
  3 |   expect<number>().type.not.toBeNumber();
> 4 |   expect<number>().type.not.toBeNumber();
    |                             ^
  5 | });
  6 | 

      at ./__typetests__/isNumber.tst.ts:4:29 ❭ is number?

Error: The source type is 'string'.

  1 | import { expect, test } from "tstyche";
  2 | test("is string?", () => {
> 3 |   expect<string>().type.not.toBeString();
    |                             ^
  4 |   expect<string>().type.not.toBeString();
  5 | });
  6 | 

      at ./__typetests__/isString.tst.ts:3:29 ❭ is string?

Error: The source type is 'string'.

  2 | test("is string?", () => {
  3 |   expect<string>().type.not.toBeString();
> 4 |   expect<string>().type.not.toBeString();
    |                             ^
  5 | });
  6 | 

      at ./__typetests__/isString.tst.ts:4:29 ❭ is string?

"
`;

exports[`'--failFast' command line option when 'false' is passed as an argument: stdout 1`] = `
"uses TypeScript <<version>>

fail ./__typetests__/isNumber.tst.ts

fail ./__typetests__/isString.tst.ts

Targets:    1 failed, 1 total
Test files: 2 failed, 2 total
Tests:      2 failed, 2 total
Assertions: 4 failed, 4 total
Duration:   <<timestamp>>

Ran all test files.
"
`;

exports[`'--failFast' command line option when 'true' is passed as an argument: stderr 1`] = `
"Error: The source type is 'number'.

  1 | import { expect, test } from "tstyche";
  2 | test("is number?", () => {
> 3 |   expect<number>().type.not.toBeNumber();
    |                             ^
  4 |   expect<number>().type.not.toBeNumber();
  5 | });
  6 | 

      at ./__typetests__/isNumber.tst.ts:3:29 ❭ is number?

"
`;

exports[`'--failFast' command line option when 'true' is passed as an argument: stdout 1`] = `
"uses TypeScript <<version>>

fail ./__typetests__/isNumber.tst.ts

Targets:    1 failed, 1 total
Test files: 1 failed, 1 total
Tests:      1 failed, 1 total
Assertions: 1 failed, 1 total
Duration:   <<timestamp>>

Ran all test files.
"
`;

exports[`'--failFast' command line option when search string is specified after the option: stderr 1`] = `
"Error: The source type is 'string'.

  1 | import { expect, test } from "tstyche";
  2 | test("is string?", () => {
> 3 |   expect<string>().type.not.toBeString();
    |                             ^
  4 |   expect<string>().type.not.toBeString();
  5 | });
  6 | 

      at ./__typetests__/isString.tst.ts:3:29 ❭ is string?

"
`;

exports[`'--failFast' command line option when search string is specified after the option: stdout 1`] = `
"uses TypeScript <<version>>

fail ./__typetests__/isString.tst.ts
  × is string?

Targets:    1 failed, 1 total
Test files: 1 failed, 1 total
Tests:      1 failed, 1 total
Assertions: 1 failed, 1 total
Duration:   <<timestamp>>

Ran test files matching 'isString'.
"
`;

exports[`'--failFast' command line option when search string is specified before the option: stderr 1`] = `
"Error: The source type is 'number'.

  1 | import { expect, test } from "tstyche";
  2 | test("is number?", () => {
> 3 |   expect<number>().type.not.toBeNumber();
    |                             ^
  4 |   expect<number>().type.not.toBeNumber();
  5 | });
  6 | 

      at ./__typetests__/isNumber.tst.ts:3:29 ❭ is number?

"
`;

exports[`'--failFast' command line option when search string is specified before the option: stdout 1`] = `
"uses TypeScript <<version>>

fail ./__typetests__/isNumber.tst.ts
  × is number?

Targets:    1 failed, 1 total
Test files: 1 failed, 1 total
Tests:      1 failed, 1 total
Assertions: 1 failed, 1 total
Duration:   <<timestamp>>

Ran test files matching 'isNumber'.
"
`;

exports[`'--failFast' command line option when the option is specified several times: stderr 1`] = `
"Error: The source type is 'number'.

  1 | import { expect, test } from "tstyche";
  2 | test("is number?", () => {
> 3 |   expect<number>().type.not.toBeNumber();
    |                             ^
  4 |   expect<number>().type.not.toBeNumber();
  5 | });
  6 | 

      at ./__typetests__/isNumber.tst.ts:3:29 ❭ is number?

Error: The source type is 'number'.

  2 | test("is number?", () => {
  3 |   expect<number>().type.not.toBeNumber();
> 4 |   expect<number>().type.not.toBeNumber();
    |                             ^
  5 | });
  6 | 

      at ./__typetests__/isNumber.tst.ts:4:29 ❭ is number?

"
`;

exports[`'--failFast' command line option when the option is specified several times: stdout 1`] = `
"uses TypeScript <<version>>

fail ./__typetests__/isNumber.tst.ts
  × is number?

Targets:    1 failed, 1 total
Test files: 1 failed, 1 total
Tests:      1 failed, 1 total
Assertions: 2 failed, 2 total
Duration:   <<timestamp>>

Ran test files matching 'isNumber'.
"
`;

exports[`'failFast' configuration file option stops running tests after the first failure: stderr 1`] = `
"Error: The source type is 'number'.

  1 | import { expect, test } from "tstyche";
  2 | test("is number?", () => {
> 3 |   expect<number>().type.not.toBeNumber();
    |                             ^
  4 |   expect<number>().type.not.toBeNumber();
  5 | });
  6 | 

      at ./__typetests__/isNumber.tst.ts:3:29 ❭ is number?

"
`;

exports[`'failFast' configuration file option stops running tests after the first failure: stdout 1`] = `
"uses TypeScript <<version>>

fail ./__typetests__/isNumber.tst.ts

Targets:    1 failed, 1 total
Test files: 1 failed, 1 total
Tests:      1 failed, 1 total
Assertions: 1 failed, 1 total
Duration:   <<timestamp>>

Ran all test files.
"
`;
